### ISSUE TITLE
An error occurred: CustomApiGatewayAccountCloudWatchRole - Modifying service token is not allowed..

### ISSUE THREAD

**Issue by yeukfei02** [serverless/serverless#9240]
**Error**

```
Serverless: Checking Stack update progress...
.......................................................
Serverless: Operation failed!
Serverless: View the full error output: https://ap-southeast-1.console.aws.amazon.com/cloudformation/home?region=ap-southeast-1#/stack/detail?stackId=arn%3Aaws%3Acloudformation%3Aap-southeast-1%3A676534811615%3Astack%2Fpython-serverless-prod%2Ff1654d80-62db-11eb-b754-02c69f0ddb0a
Serverless: Publishing service to the Serverless Dashboard...
Serverless: Successfully published your service to the Serverless Dashboard: https://app.serverless.com/yeukfei02/apps/python-serverless/python-serverless/prod/ap-southeast-1
 
  Serverless Error ---------------------------------------
 
  An error occurred: CustomApiGatewayAccountCloudWatchRole - Modifying service token is not allowed..
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              12.21.0
     Framework Version:         2.21.1 (local)
     Plugin Version:            4.4.2
     SDK Version:               2.3.2
     Components Version:        3.6.0
```

**serverless.yml**

```yaml
org: yeukfei02
app: python-serverless
service: python-serverless

frameworkVersion: "2"
useDotenv: true

provider:
  name: aws
  runtime: python3.8
  stage: prod
  region: ap-southeast-1
  apiName: python-serverless-prod
  tracing:
    apiGateway: true
    lambda: true
  environment:
    PYTHON_ENV: production
    MY_AWS_ACCESS_KEY: ${param:MY_AWS_ACCESS_KEY}
    MY_AWS_SECRET_ACCESS_KEY: ${param:MY_AWS_SECRET_ACCESS_KEY}
    JWT_SECRET: ${param:JWT_SECRET}
  iamRoleStatements:
    - Effect: "Allow"
      Action:
        - "xray:PutTraceSegments"
        - "xray:PutTelemetryRecords"
      Resource:
        - "*"
    - Effect: "Allow"
      Action:
        - "dynamodb:Query"
        - "dynamodb:Scan"
        - "dynamodb:GetItem"
        - "dynamodb:PutItem"
        - "dynamodb:UpdateItem"
        - "dynamodb:DeleteItem"
        - "dynamodb:DescribeTable"
      Resource:
        - "*"
  apiGateway:
    shouldStartNameWithService: true
    description: python serverless prod
  lambdaHashingVersion: 20201221

functions:
  main:
    handler: src/main/handler.main
    events:
      - http:
          path: /
          method: get
          cors: true

  signup:
    handler: src/signup/handler.signup
    events:
      - http:
          path: /signup
          method: post
          cors: true

  login:
    handler: src/login/handler.login
    events:
      - http:
          path: /login
          method: post
          cors: true

  authorize:
    handler: src/authorize/handler.authorize

  createNotes:
    handler: src/createNotes/handler.createNotes
    events:
      - http:
          path: /notes/create-notes
          method: post
          cors: true
          authorizer:
            name: authorize

  getNotes:
    handler: src/getNotes/handler.getNotes
    events:
      - http:
          path: /notes
          method: get
          cors: true
          authorizer:
            name: authorize

  getNoteById:
    handler: src/getNoteById/handler.getNoteById
    events:
      - http:
          path: /notes/{id}
          method: get
          cors: true
          authorizer:
            name: authorize

  updateNoteById:
    handler: src/updateNoteById/handler.updateNoteById
    events:
      - http:
          path: /notes/{id}
          method: patch
          cors: true
          authorizer:
            name: authorize

  deleteNoteById:
    handler: src/deleteNoteById/handler.deleteNoteById
    events:
      - http:
          path: /notes/{id}
          method: delete
          cors: true
          authorizer:
            name: authorize

plugins:
  - serverless-python-requirements
  - serverless-prune-plugin
  - serverless-offline
  # - serverless-plugin-split-stacks

custom:
  pythonRequirements:
    dockerizePip: true
    pythonBin: python3
  prune:
    automatic: true
    number: 3
```

I am using github actions to deploy python serverless to aws lambda...have this error...
nodejs lambda func use with `serverless-plugin-split-stacks` also same.. anyway can fix this?

Thanks a lot.

**Comment by medoror** [comment]
I am getting the same error output. I was able to successfully deploy my application until I added an [aws lambda layer](https://www.serverless.com/blog/publish-aws-lambda-layers-serverless-framework) to my deployment

```
Serverless: Checking Stack update progress...
..............
Serverless: Operation failed!
Serverless: View the full error output: https://us-west-2.console.aws.amazon.com/cloudformation/home?region=us-west-2#/stack/detail?stackId=arn%3Aaws%3Acloudformation%3Aus-west-2%3A703485809922%3Astack%2Faudio-clipper-slackbot-dev%2Fc1b64e30-7010-11eb-8474-065e4333b4d9
Serverless: Publishing service to the Serverless Dashboard...
Serverless: Successfully published your service to the Serverless Dashboard: https://app.serverless.com/edoror/apps/audio-clipper/audio-clipper-slackbot/dev/us-west-2

 Serverless Error ----------------------------------------

  An error occurred: CustomApiGatewayAccountCloudWatchRole - Modifying service token is not allowed..

  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com

  Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              13.12.0
     Framework Version:         2.33.1
     Plugin Version:            4.5.3
     SDK Version:               4.2.2
     Components Version:        3.8.1
```

Here [serverless.yaml](https://github.com/medoror/audio_clipper_slackbot/blob/master/serverless.yml)

**Comment by yeukfei02** [comment]
@medoror 

if I don't want to use aws lambda layer then how?

**Comment by medoror** [comment]
hey @yeukfei02 I am no longer getting this error.  I found [this page](https://forum.serverless.com/t/modifying-service-token-is-not-allowed/12404)

I am not exactly sure what is going on but my app is now deploying after I added the `stage` flag to my deploy command
`npx serverless deploy --stage dev`

**Comment by yeukfei02** [comment]
@medoror 

are u using any cicd? In my case I am using github actions

https://github.com/serverless/github-action

**Comment by medoror** [comment]
@yeukfei02 I am not, just a regular app deploy

**Comment by yeukfei02** [comment]
@medoror 

thanks for the answer, I am able to deploy without error

**Comment by rodrigogs** [comment]
Why was it closed. There's nothing to understand about this issue here.

Why do we need to use `--stage` flag now? Are you calling this an improvement? It's just breaking stuff...

Serverless Framework have been accumulating issues while you're trying do adapt to version 3. Some things works well, other are pretty broken, and the community has almost no info about those new issues.

**Comment by yeukfei02** [comment]
@rodrigogs 

sure lets reopen this again..i also don't know why need to specify `--stage`..in `serverless.yml` already define

**Comment by rodrigogs** [comment]
Nice. I really wish I could adapt my services to v3, but the documentation about the migration is soooooooooo superficial.

**Comment by yeukfei02** [comment]
I feel using serverless framework still better to deploy from local machine..rather than rely on the cicd..
sometimes the environment is not same as local...

**Comment by pgrzesik** [comment]
Hello @yeukfei02 - would you be able to provide a minimal reproducible example of that issue where it doesn't work if `--stage` is not passed as a CLI param? I tried to reproduce it locally but no luck and a minimal reproducible example would help tremendously in debugging that issue. Thanks in advance.

@rodrigogs Could you share a bit more about "other are pretty broken" changes that you mention? What could be improved in your opinion? What is not clear?

**Comment by yeukfei02** [comment]
@pgrzesik 

the error message I posted in the first post

This is the error log from github actions
```
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
...............................
Serverless: Operation failed!
Serverless: View the full error output: https://ap-southeast-1.console.aws.amazon.com/cloudformation/home?region=ap-southeast-1#/stack/detail?stackId=arn%3Aaws%3Acloudformation%3Aap-southeast-1%3A676534811615%3Astack%2Frandom-user-api-serverless-prod%2F5b9cee90-3e71-11eb-946c-02227c3df95e
Serverless: Publishing service to the Serverless Dashboard...
Serverless: Successfully published your service to the Serverless Dashboard: https://app.serverless.com/yeukfei02/apps/random-user-api-serverless/random-user-api-serverless/prod/ap-southeast-1
 
  Serverless Error ---------------------------------------
 
  An error occurred: CustomApiGatewayAccountCloudWatchRole - Modifying service token is not allowed..
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              14.16.0
     Framework Version:         2.21.1 (local)
     Plugin Version:            4.4.2
     SDK Version:               2.3.2
     Components Version:        3.5.1
```

**Comment by pgrzesik** [comment]
Hello @yeukfei02 - I saw that, but that doesn't help reproducing the issue - could you provide a minimal example of `serverless.yml` where that problem happens? I've tried to reproduce with a simple example but I didn't manage to hit the same error.

**Comment by yeukfei02** [comment]
@pgrzesik 

this is one of my project `serverless.yml`

```
org: yeukfei05
app: random-user-api-serverless
service: random-user-api-serverless

frameworkVersion: '2'
useDotenv: true

provider:
  name: aws
  runtime: nodejs12.x
  stage: prod
  region: ap-southeast-1
  apiName: random-user-api-serverless-prod
  tracing:
    apiGateway: true
    lambda: true
  environment:
    NODE_ENV: production
    MY_AWS_ACCESS_KEY: ${param:MY_AWS_ACCESS_KEY}
    MY_AWS_SECRET_ACCESS_KEY: ${param:MY_AWS_SECRET_ACCESS_KEY}
    JWT_SECRET: ${param:JWT_SECRET}
  iam:
    role:
      statements:
        - Effect: 'Allow'
          Action:
            - 'xray:PutTraceSegments'
            - 'xray:PutTelemetryRecords'
          Resource:
            - '*'
  apiGateway:
    shouldStartNameWithService: true
    description: random user api serverless prod
  lambdaHashingVersion: 20201221

functions:
  main:
    handler: src/main/handler.getMain
    events:
      - http:
          path: /api
          method: get
          cors: true

  signup:
    handler: src/signup/handler.signup
    events:
      - http:
          path: /api/user/signup
          method: post
          cors: true

  login:
    handler: src/login/handler.login
    events:
      - http:
          path: /api/user/login
          method: post
          cors: true

  authorize:
    handler: src/authorize/handler.authorize

  createRandomUser:
    handler: src/createRandomUser/handler.createRandomUser
    events:
      - http:
          path: /api/random-user
          method: post
          cors: true
          authorizer:
            name: authorize

  getRandomUser:
    handler: src/getRandomUser/handler.getRandomUser
    events:
      - http:
          path: /api/random-user
          method: get
          cors: true
          authorizer:
            name: authorize

  getRandomUserById:
    handler: src/getRandomUserById/handler.getRandomUserById
    events:
      - http:
          path: /api/random-user/{id}
          method: get
          cors: true
          authorizer:
            name: authorize

  updateRandomUserById:
    handler: src/updateRandomUserById/handler.updateRandomUserById
    events:
      - http:
          path: /api/random-user/{id}
          method: put
          cors: true
          authorizer:
            name: authorize

  deleteRandomUserById:
    handler: src/deleteRandomUserById/handler.deleteRandomUserById
    events:
      - http:
          path: /api/random-user/{id}
          method: delete
          cors: true
          authorizer:
            name: authorize

plugins:
  - serverless-prune-plugin
  - serverless-plugin-typescript
  - serverless-offline
  - serverless-plugin-split-stacks

custom:
  prune:
    automatic: true
    number: 3
```

**Comment by pgrzesik** [comment]
Thank you @yeukfei02 - I saw that, but it's a bit extensive example with multiple plugins used which makes it harder to actually nail down specific issue if such exists. Would you be able to trim it down to only essentials, ideally without external plugins so we have a minimal reproducible example of the issue? E.g. only with only one simple function and without any plugins and additional unrelated settings. It would be really helpful

**Comment by pgrzesik** [comment]
Additionally, could you please check if that problem still happens for you after upgrading to latest Framework version? I see that you're a bit behind on versions

**Comment by joelcogen** [comment]
I have a similar issue, this is really weird.

serverless.yml:

```
frameworkVersion: '2'

provider:
  name: aws
  runtime: nodejs14.x
  lambdaHashingVersion: 20201221
  region: eu-central-1
  stage: staging
```

- `sls deploy` works
- `sls deploy --stage staging` doesn't work
- CI/CD with **staging** stage  doesn't work

**Comment by yeukfei02** [comment]
@joelcogen 

hello, what is your `serverless` version? for my case, update sls version to latest and go sls dashboard add a new accesskey/role arn to provider looks ok

**Comment by joelcogen** [comment]
```
Your Environment Information ---------------------------
     Operating System:          darwin
     Node Version:              15.7.0
     Framework Version:         2.34.0 (local)
     Plugin Version:            4.5.3
     SDK Version:               4.2.2
     Components Version:        3.8.2
```

Seems everything is already at the latest version. I created a new provider, same issue.

**Comment by joelcogen** [comment]
@pgrzesik I don't have time to reproduce in a brand new app right now but the way I understand it:

1. create a new app on aws
2. set a `provider.stage` in serverless.yml
3. `sls deploy`
4. `sls deploy --stage the_same_stage` should trigger the error

**Comment by joelcogen** [comment]
Here is another maybe interesting error.

My app is called **lansky** and the problem stage is **staging**.

I tried changing `provider.stage` in serverless.yml to **test-bug-stage**.

`sls deploy`:

```
An error occurred: CustomDashresourceDashapigwDashcwDashroleLambdaFunction - lansky-undefined-custom-resource-apigw-cw-role already exists in stack arn:aws:cloudformation:eu-central-1:XXXXX:stack/lansky-staging/XXX-XXX-XXX.
```

(ids replaced with XX)

The fact it is trying to create **lansky-undefined-…** makes me think that somewhere the stage isn't read correctly from `provider.stage`, but with no prior knowledge of the code I'm stuck here and not sure of the impact

**Comment by yeukfei02** [comment]
@joelcogen 

did u encounter lambda function if delete from aws console show access denied?
or did u try `sls remove`..is it remove all apigateway, lambda, s3, iamrole, cloudwatch, cloudformation?
`lansky-undefined-custom-resource-apigw-cw-role` lambda func should be remove first, then u can do `sls deploy` again

**Comment by yeukfei02** [comment]
@joelcogen 

i face this issue in #9242

**Comment by pgrzesik** [comment]
Hello @joelcogen and @yeukfei02 I've tried to reproduce that issue locally and I've tried multiple different versions of the Framework, Node, different configs but I'm not able to reproduce it. @joelcogen would you be able by any chance to provide a minimal reproducible example of that issue? It would make reproducing it much easier

**Comment by pgrzesik** [comment]
I didn't manage to reproduce locally, but I've spotted a problematic piece of code that could cause that and proposed a potential fix.

**Comment by joelcogen** [comment]
Not sure what changed but now both `sls deploy` and `sls deploy --stage staging` give the error

**Comment by pgrzesik** [comment]
@joelcogen Could you provide a minimal reproducible example?

**Comment by joelcogen** [comment]
@yeukfei02 Thanks, removing and recreating fixed the problem

**Comment by Zambonilli** [comment]
Hit this same error because we tried to add a ManagedPolicy to the IamRoleCustomResourcesLambdaExecution resource via an extension after originally being deployed with no such restriction. I think IAM is actually rejecting the change and cloudformation and serverless just pass the error through. This was with the latest version of serverless, 3.7.1

Like others have said, manually creating a role and using that for the next serverless deploy removes the role from cloudformation and then you can line up the following release to go back to serverless managing the role. In our case, setting aws provider's `provider.logs.restApi.role` to our handcrafted role and `provider.logs.restApi.roleManagedExternally` to true worked. Then a subsequent deploy we removed our handcrafted role and removed setting these fields so serverless can take over management with new objects again.

**Comment by bjornhansen** [comment]
I'm seeing this error in version 3.34.0 and unfortunately, none of the things mentioned in this thread have resolved it.

**Version information:**
```
Framework Core: 3.34.0 (local) 3.33.0 (global)
Plugin: 6.2.3
SDK: 4.3.2
```

**What I'm doing:**
Trying to deploy a stage named `prod` after deploying the default `dev` stage for an app. The stages are in the same provider. I set up a project, then ran `sls deploy`, which was successful, followed by `sls deploy --stage prod` which is where I encountered the error.

**The message I'm getting:**
```
Error:
UPDATE_FAILED: CustomApiGatewayAccountCloudWatchRole (Custom::ApiGatewayAccountRole)
Modifying service token is not allowed.
```

**What I've tried:**
1. Using different stage names. E.g. `main`.
2. Using a custom `provider.logs.restApi.role` and setting `provider.logs.restApi.roleManagedExternally` to true as suggested by @Zambonilli.
3. Deploying the "prod" stage to a separate "service" in Serverless.

Whatever I do, anything but the default `dev` stage results in that error. Has this still not been resolved?
